drivers/adc/mcp356x/mcp356x.c

```diff
@@ -0,0 +1,542 @@
+
+    uint8_t mux_reg = data->channel_map[i].mux_reg;
+    if (mux_reg != data->mux) {
+      (void)mcp_write_reg_8(config, MCP_Reg_MUX, mux_reg);
+      data->mux = mux_reg;
+
+      // Wait for result to come in after ADC reset because of MUX change
+      // right one
+      k_usleep(10);
+    }
+
+    uint32_t val = 12345;
+    res = mcp_read_reg_24(config, MCP_Reg_ADCDATA, &val);
+    if (res != 0) {
+      return res;
+    }
+    sample_array[sample_index] = val; // TODO FIX THIS
+    sample_index++;
+  }
+
+  return 0;
+}
+
+int mcp356x_channel_setup(const struct device *dev,
+                          const struct adc_channel_cfg *channel_cfg) {
+  if (!device_is_ready(dev)) {
+    LOG_ERR("Base mcp356x device is not setup. Can not setup channel");
+    return -ENODEV;
+  }
+  struct mcp356x_data *data = dev->data;
+
+  if (channel_cfg->channel_id > 8) {
+    LOG_ERR(
+        "register/channel %d of %s invalid. Any mcp356x device supports max 4 "
+        "differential channels or 8 single ended channels. You probably "
+        "didn't mean to have this many channels",
+        channel_cfg->channel_id, dev->name);
+    return -ENODEV;
+  }
+
+  uint8_t mux_vin_p = channel_cfg->input_positive;
+  if (mux_vin_p > 15) {
+    LOG_ERR("zephyr,positive invalid for %s channel %d.", dev->name,
+            channel_cfg->channel_id);
+  }
+  uint8_t mux_vin_m = channel_cfg->input_negative;
+  if (mux_vin_m > 15) {
+    LOG_ERR("zephyr,negative invalid for %s channel %d.", dev->name,
+            channel_cfg->channel_id);
+  }
+
+  uint8_t mux_reg = (mux_vin_p << 4) | mux_vin_m;
+  bool differential = channel_cfg->differential;
+  uint8_t gain_bits = 0;
+  switch (channel_cfg->gain) {
+  case ADC_GAIN_1_3:
+    gain_bits = 0b000;
+    break;
+  case ADC_GAIN_1:
+    gain_bits = 0b001;
+    break;
+  case ADC_GAIN_2:
+    gain_bits = 0b010;
+    break;
+  case ADC_GAIN_4:
+    gain_bits = 0b011;
+    break;
+  case ADC_GAIN_8:
+    gain_bits = 0b100;
+    break;
+  case ADC_GAIN_16:
+    gain_bits = 0b101;
+    break;
+  case ADC_GAIN_32:
+    gain_bits = 0b110;
+    break;
+  case ADC_GAIN_64:
+    gain_bits = 0b111;
+    break;
+  default:
+    LOG_ERR("unsupported channel gain '%d'", channel_cfg->gain);
+    return -ENOTSUP;
+  }
+
+  enum adc_reference ref = channel_cfg->reference;
+
+  // VREF Selection
+  uint8_t vref_sel_bits = 0b0;
+  switch (ref) {
+  case ADC_REF_INTERNAL:
+    vref_sel_bits = 0b1;
+    break;
+  case ADC_REF_EXTERNAL0:
+    vref_sel_bits = 0b0;
+    break;
+  default:
+    LOG_ERR("unsupported reference voltage '%d'", ref);
+    return -ENOTSUP;
+  }
+
+  struct channel_map_entry entry = {.mux_reg = mux_reg,
+                                    .differential = differential,
+                                    .gain_bits = gain_bits,
+                                    .reference_bits = ref};
+
+  data->channel_map[channel_cfg->channel_id] = entry;
+
+  return 0;
+}
+
+static const struct adc_driver_api mcp356x_api = {
```

Comment:
Just be aware that this may cause problems if we choose to use multiple devices with different configs. May not be a problem for our future use cases though.